### ISSUE TITLE
Doesn’t work with Yarn PnP

### ISSUE THREAD

**Issue by lensbart** [remix-run/remix#4346]
### What version of Remix are you using?

1.7.2

#### Enable Yarn PnP
According to [the official instructions](https://yarnpkg.com/getting-started/install). If you don’t have Node.js >= 16.10 installed, run `npm i -g corepack` first.

```bash
corepack enable
yarn init -2
```

#### Install Remix

```bash
mkdir packages
cd packages
yarn create remix
```

and answer the questions in the CLI prompt.

#### First problem
`cd` into your newly created app, and run `yarn dev`. You’ll see the following error:

> Error: Could not locate @remix-run/serve. Please verify you have it installed to use the dev command.

I submitted a PR to fix this problem [here](https://github.com/remix-run/remix/pull/3566), but it was declined, saying that the fix will be redundant once the switch is made to `@remix-run/router`. That was in June. I still think it would be useful if Remix could work out-of-the-box for Yarn PnP users in the meantime.

A workaround for this error is to add the following to the `.yarnrc.yml` file of the consuming project:

```yml
# .yarnrc.yml
packageExtensions:
  react-router-dom@*:
    dependencies:
      react: '*'
      react-dom: '*'
```

I don’t think this workaround obviates a fix, because the actual error is currently hidden:

> Error: react-router-dom tried to access react (a peer dependency) but it isn't provided by its ancestors; this makes the require call ambiguous and unsound.

Thus, I’d also suggest adding `console.log(err)` between [these two lines](https://github.com/remix-run/remix/blob/e260b2ba153271fed2e2c569abdfecdccf241625/packages/remix-dev/cli/commands.ts#L273-L274).

#### Second problem

After adding the lines to `.yarnrc.yml` and running `yarn` again, the project compiles. However, Remix will incorrectly warn that installed dependencies are not installed:

> The path "@remix-run/react" is imported in src/root.tsx but "@remix-run/react" was not found in your node_modules. Did you forget to install it?

I created a PR to fix this problem [here](https://github.com/remix-run/remix/pull/3599). [edit: merged now]

#### Third problem

By default, Remix configures an alias `~` for more convenient imports. These don’t work with `@yarnpkg/esbuild-plugin-pnp`:

>  [ERROR] [plugin @yarnpkg/esbuild-plugin-pnp] Your application tried to access ~, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.
> 
> Required package: ~ (via "~/something")
> Required by: /some-folder/packages/app/src/
> 
> 
>     pnp:/some-folder/packages/app/src/someFile.ts:3:26:
>       3 │ import { something } from '~/something'
>         ╵ 

`@yarnpkg/esbuild-plugin-pnp` [isn’t needed anymore](https://github.com/yarnpkg/berry/tree/master/packages/esbuild-plugin-pnp) with esbuild 0.15. Remix is still on 0.14. I’m not sure if the error will automatically disappear when upgrading esbuild. But I’m hopeful, because esbuild does support `paths` in `tsconfig.json` files.

Would it be possible to update the esbuild dependency soon? [edit: [PR here](https://github.com/remix-run/remix/pull/4301)]

Thanks!

**Fourth problem**
Import assertions don’t seem to work.
```ts
export const loader = async ({ request }: LoaderArgs) => {
  const language = 'en' // actually get these from `request.headers.get('accept-language')`

  const messages = await import(`../public/lang/${language}.json`, {
    assert: { type: 'json' },
  })

  return { messages }
}
```
gives the following error:
> TypeError: [ERR_IMPORT_ASSERTION_TYPE_MISSING]: Module "file:///some-directory/packages/app/public/lang/en.json" needs an import assertion of type "json"
>    at load$1 (file:///some-directory/.pnp.loader.mjs:1455:17)
>    at nextLoad (node:internal/modules/esm/loader:163:28)
>    at ESMLoader.load (node:internal/modules/esm/loader:605:26)
>    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:22)
>    at new ModuleJob (node:internal/modules/esm/module_job:63:26)
>    at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:480:17)
>    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34)
>    at async Promise.all (index 0)
>    at ESMLoader.import (node:internal/modules/esm/loader:530:24)
>    at importModuleDynamicallyWrapper (node:internal/vm/module:438:15)

### Expected Behavior

Remix works out-of-the-box with Yarn PnP

### Actual Behavior

Remix doesn’t work out-of-the-box with Yarn PnP, see above.

**Comment by electrolyte-orb** [comment]
Are you pleased to provide me with a GitHub Repo for this?
PS: Lately I have been working on many Yarn PNP projects and there are quirks here and there. [Take a look](https://github.com/electrolyte-orb/firebase-nextjs/blob/69edd25fd17844e35746f7957de80abb3e1d41ba/.circleci/config.yml#L18)

**Comment by lensbart** [comment]
Isn’t the fix already in the linked PR’s? I’m currently using `0.0.0-experimental-aed5bfbf7`, which includes the esbuild version bump and allows me to use Remix, albeit with warnings.

@mcansh Any chance we could get [this](https://github.com/remix-run/remix/pull/4301) merged, given that it’s tested & approved?

**Comment by electrolyte-orb** [comment]
Try to deploy that

**Comment by MichaelDeBoey** [comment]
Closed by #4301
